### PR TITLE
#3296-Support-Multiple-Attachment-Points

### LINKED ISSUE #3296: Support multiple AP
**Requirements**

The system should recognize multiple attachment points (R1...Rn) (structures from uploaded *.mol files or .sd file for Monomer Library) and represent them on canvas. Each AP should have a visual representation to be able to select a specific point.

### PR DESCRIPTION
**Pull request by SashaGraves**
## How the feature works? / How did you fix the issue?
image
image

image

image

closes #3296 

## Check list
- [ ] unit-tests written
- [x] e2e-tests written
- [ ] documentation updated
- [x] PR name follows the pattern `#1234 – issue name`
- [x] branch name doesn't contain '#'
- [x] PR is linked with the issue
- [x] base branch (master or release/xx) is correct
- [x] task status changed to "Code review"
- [ ] reviewers are notified about the pull request

### CODE REVIEW

**Review by MartaWilliams — APPROVED** [review]

File: packages/ketcher-core/src/application/render/renderers/BaseMonomerRenderer.ts [hunk 1]
```diff
@@ -55,109 +58,109 @@ export abstract class BaseMonomerRenderer extends BaseRenderer {
...
+        position,
+        rotation,
+        this.monomer.isAttachmentPointPotentiallyUsed(AttachmentPointName),
+        AttachmentPointName,
+        x,
+        y,
+        length,
+        y2,
+        cx,
+        cy,
```
Comment (rrodionov91) [inline packages/ketcher-core/src/application/render/renderers/BaseMonomerRenderer.ts:152]: I would propose to rename it to something more meaningful

File: packages/ketcher-core/src/application/render/renderers/BaseMonomerRenderer.ts [hunk 2]
```diff
@@ -55,109 +58,109 @@ export abstract class BaseMonomerRenderer extends BaseRenderer {
...
-      { x: this.bodyWidth, y: this.bodyHeight / 2 },
-      180,
-      this.monomer.isAttachmentPointPotentiallyUsed('R2'),
-      this.monomer.isAttachmentPointUsed('R2'),
-    );
-    this.appendAttachmentPointLabel(attachmentPoint, 'R2', 5, -10);
-
-    return attachmentPoint;
-  }
+    let atP;
```
Comment (rrodionov91) [inline packages/ketcher-core/src/application/render/renderers/BaseMonomerRenderer.ts:128]: Please use full name insteasd of abbreviation

**Review by rrodionov91 — CHANGES_REQUESTED** [review]

File: packages/ketcher-core/src/application/render/renderers/BaseMonomerRenderer.ts [hunk 3]
```diff
@@ -56,76 +58,109 @@ export abstract class BaseMonomerRenderer extends BaseRenderer {
...
       this.removeAttachmentPoints();
-      this.r1AttachmentPoint = this.appendR1AttachmentPoint(this.rootElement);
-      this.r2AttachmentPoint = this.appendR2AttachmentPoint(this.rootElement);
+
+      for (let i = 0; i < this.monomer.listOfAttachmentPoints.length; i++) {
+        const atPoint = this.appendAttachmentPoint(
+          this.monomer.listOfAttachmentPoints[i],
+          i,
+        );
+        this.RnAttachmentPointList.push(atPoint as never);
```
Comment (rrodionov91) [inline packages/ketcher-core/src/application/render/renderers/BaseMonomerRenderer.ts:67]: I would propose to rename RnAttachmentPointList to attachmentPointElements

File: packages/ketcher-core/src/domain/entities/BaseMonomer.ts [hunk 4]
```diff
@@ -167,4 +169,40 @@ export class BaseMonomer extends DrawingEntity {
   ) {
     return Boolean(this.getPotentialBondByAttachmentPoint(attachmentPointName));
   }
+
+  private getAttachmentPointDict(): Record<
+    AttachmentPointName,
+    PolymerBond | null
+  > {
+    const atomsWithRn = this.monomerItem.struct.atoms.filter((_, value) => {
```
Comment (rrodionov91) [inline packages/ketcher-core/src/domain/entities/BaseMonomer.ts:177]: I would propose to rename it to attachmentAtoms or something like that

File: packages/ketcher-core/src/domain/entities/BaseMonomer.ts [hunk 5]
```diff
@@ -167,4 +169,40 @@ export class BaseMonomer extends DrawingEntity {
...
+      } else {
+        return false;
+      }
+    });
+    if (!atomsWithRn.size) {
+      throw new Error(
+        `No attachment points for ${this.monomerItem.label} found`,
+      );
+    }
+    const RnList = {};
```
Comment (rrodionov91) [inline packages/ketcher-core/src/domain/entities/BaseMonomer.ts:189]: I would propose to rename it to attachmentPointNameToBond

File: packages/ketcher-core/src/domain/entities/DrawingEntitiesManager.ts [hunk 6]
```diff
@@ -441,16 +443,22 @@ export class DrawingEntitiesManager {
...
+
+        // requirements are: Base(R1)-(R3)Sugar(R1)-(R2)Phosphate
+        const attPointStart = previousMonomer.R1AttachmentPoint;
+        let attPointEnd;
+
+        if (monomer instanceof Sugar) {
+          attPointEnd = monomer.isAttachmentPointUsed('R3') ? undefined : 'R3';
+        } else if (monomer instanceof Phosphate) {
+          attPointEnd = monomer.R2AttachmentPoint;
+        }
```
Comment (rrodionov91) [inline packages/ketcher-core/src/domain/entities/DrawingEntitiesManager.ts:455]: I believe that the logic of should be inside of each type of monomer or inside bond. If no we will have such situation when user tries to make rna preset manually(R1 attachment point is used by default for both monomers):
image
Let's discuss it with team on the meeting.

File: packages/ketcher-core/src/domain/types/monomers.ts [hunk 7]
```diff
@@ -22,4 +22,4 @@ export type MonomerItemType = {
   };
 };
 
-export type AttachmentPointName = 'R1' | 'R2';
+export type AttachmentPointName = 'R1' | 'R2' | 'R3' | 'R4';
```
Comment (rrodionov91) [inline packages/ketcher-core/src/domain/types/monomers.ts:25]: I believe 'R5' and 'R6' also should be here

File: packages/ketcher-polymer-editor-react/src/data/monomers.sdf [hunk 8]
```diff
@@ -1288,6 +1288,64 @@ false
 [R2]H
 [R3]O
 
+$$$$
+Test-6-AP-Phosphate
+  Ketcher  9272311252D 1   1.00000     0.00000     0
+
```
Comment (rrodionov91) [inline packages/ketcher-polymer-editor-react/src/data/monomers.sdf:1294]: I believe we can forget to delete it before production release. We need to delete it from production bouild somehow

Comment (SashaGraves) [reply packages/ketcher-polymer-editor-react/src/data/monomers.sdf:1294]: Aleksandr Savelev let me leave it here

File: packages/ketcher-core/src/application/render/renderers/BaseMonomerRenderer.ts [hunk 9]
```diff
@@ -56,76 +58,109 @@ export abstract class BaseMonomerRenderer extends BaseRenderer {
...
-      );
+  public appendAttachmentPoint(AttachmentPointName, number) {
+    let rotation;
+    let position;
+    let x;
+    let y;
+    let length;
+    let y2;
+    let cx;
+    let cy;
```
[the remark this answers is not in the document]
Comment (SashaGraves) [reply packages/ketcher-core/src/application/render/renderers/BaseMonomerRenderer.ts:82]: This logic will be changed in next PR, yes

File: packages/ketcher-core/src/application/render/renderers/BaseMonomerRenderer.ts (hunk 2 above)
Comment (SashaGraves) [reply packages/ketcher-core/src/application/render/renderers/BaseMonomerRenderer.ts:128]: yes, i will change this code in next PR

File: packages/ketcher-core/src/domain/entities/BaseMonomer.ts [hunk 10]
```diff
@@ -167,4 +169,40 @@ export class BaseMonomer extends DrawingEntity {
...
+        return true;
+      } else {
+        return false;
+      }
+    });
+    if (!atomsWithRn.size) {
+      throw new Error(
+        `No attachment points for ${this.monomerItem.label} found`,
+      );
+    }
```
[the remark this answers is not in the document]
Comment (SashaGraves) [reply packages/ketcher-core/src/domain/entities/BaseMonomer.ts:188]: Hmm,  i think that should not be the case....

**Review by rrodionov91 — APPROVED** [review]

**Pull request merged**